src/audio/rtnr/rtnr.c

```diff
@@ -766,96 +735,95 @@ static int rtnr_copy(struct comp_dev *dev)
 
 	comp_dbg(dev, "rtnr_copy()");
 
-	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
-
 	/* put empty data into output queue*/
-	RTKMA_API_First_Copy(cd->rtk_agl, cd->source_rate,
-			     audio_stream_get_channels(&source->stream));
+	RTKMA_API_First_Copy(cd->rtk_agl, cd->source_rate, audio_stream_get_channels(source));
 
-	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	if (!frames)
+		return 0;
 
+	comp_dbg(dev, "rtnr_copy() frames = %d", frames);
 	if (cd->process_enable) {
-		frames = audio_stream_avail_frames(&source->stream, &sink->stream);
-
-		/* Process integer multiple of RTNR internal block length */
-		frames = frames & ~RTNR_BLK_LENGTH_MASK;
+		/* Run processing function */
 
-		comp_dbg(dev, "rtnr_copy() source->id: %d, frames = %d", source->id, frames);
+		/* copy required data from sof audio stream to RTNR audio stream */
+		rtnr_copy_from_sof_stream(sources_stream[0], source);
+		rtnr_copy_from_sof_stream(sink_stream, sink);
 
-		if (frames) {
-			source_bytes = frames * audio_stream_frame_bytes(&source->stream);
-			sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);
-
-			buffer_stream_invalidate(source, source_bytes);
-
-			/* Run processing function */
+		/*
+		 * Processing function uses an array of pointers to source streams
+		 * as parameter.
+		 */
+		cd->rtnr_func(mod, (struct audio_stream_rtnr **)&sources_stream,
```

Comment:
I know this is copied from the current version, but I think, that if you use `sources_stream` instead of `&sources_stream` you get the same result, but a better readability and you won't need the type-cast. In all locations of course